bundles/org.openhab.binding.tuya/src/main/java/org/openhab/binding/tuya/internal/local/handlers/TuyaDecoder.java

```diff
@@ -0,0 +1,219 @@
+
+import java.nio.ByteBuffer;
+import java.util.Arrays;
+import java.util.Base64;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.openhab.binding.tuya.internal.local.CommandType;
+import org.openhab.binding.tuya.internal.local.MessageWrapper;
+import org.openhab.binding.tuya.internal.local.ProtocolVersion;
+import org.openhab.binding.tuya.internal.local.dto.DiscoveryMessage;
+import org.openhab.binding.tuya.internal.local.dto.TcpStatusPayload;
+import org.openhab.binding.tuya.internal.util.CryptoUtil;
+import org.openhab.core.util.HexUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.Gson;
+import com.google.gson.JsonSyntaxException;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.handler.codec.ByteToMessageDecoder;
+
+/**
+ * The {@link TuyaDecoder} is a Netty Decoder for encoding Tuya Local messages
+ *
+ * Parts of this code are inspired by the TuyAPI project (see notice file)
+ *
+ * @author Jan N. Klug - Initial contribution
+ */
+@NonNullByDefault
+public class TuyaDecoder extends ByteToMessageDecoder {
+    private final Logger logger = LoggerFactory.getLogger(TuyaDecoder.class);
+
+    private final Gson gson;
+
+    public TuyaDecoder(Gson gson) {
+        this.gson = gson;
+    }
+
+    @Override
+    public void decode(@NonNullByDefault({}) ChannelHandlerContext ctx, @NonNullByDefault({}) ByteBuf in,
+            @NonNullByDefault({}) List<Object> out) throws Exception {
+        if (in.readableBytes() < 24) {
+            // minimum packet size is 16 bytes header + 8 bytes suffix
+            return;
+        }
+
+        if (!ctx.channel().hasAttr(DEVICE_ID_ATTR) || !ctx.channel().hasAttr(PROTOCOL_ATTR)
+                || !ctx.channel().hasAttr(SESSION_KEY_ATTR)) {
+            logger.warn(
+                    "{}: Failed to retrieve deviceId, protocol or sessionKey from ChannelHandlerContext. This is a bug.",
+                    Objects.requireNonNullElse(ctx.channel().remoteAddress(), ""));
+            return;
+        }
+        String deviceId = ctx.channel().attr(DEVICE_ID_ATTR).get();
+        ProtocolVersion protocol = ctx.channel().attr(PROTOCOL_ATTR).get();
+        byte[] sessionKey = ctx.channel().attr(SESSION_KEY_ATTR).get();
+
+        // we need to take a copy first so the buffer stays intact if we exit early
+        ByteBuf inCopy = in.copy();
+        byte[] bytes = new byte[inCopy.readableBytes()];
+        inCopy.readBytes(bytes);
+        inCopy.release();
+
+        if (logger.isTraceEnabled()) {
+            logger.trace("{}{}: Received encoded '{}'", deviceId,
+                    Objects.requireNonNullElse(ctx.channel().remoteAddress(), ""), HexUtils.bytesToHex(bytes));
+        }
+
+        ByteBuffer buffer = ByteBuffer.wrap(bytes);
+        int prefix = buffer.getInt();
+        int sequenceNumber = buffer.getInt();
```

Comment:
Compiler warning:
```
[WARNING] /var/jenkins_home/workspace/PR-openHAB-Addons/bundles/org.openhab.binding.tuya/src/main/java/org/openhab/binding/tuya/internal/local/handlers/TuyaDecoder.java:[99,13] The value of the local variable sequenceNumber is not used
```

Reply:
It's also possible to remove the assignment, but the method call is necessary to correctly live the pointer within the buffer.